client/rest-high-level/src/main/java/org/elasticsearch/client/AsyncSearchRequestConverters.java

```diff
@@ -0,0 +1,74 @@
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import org.apache.http.client.methods.HttpPost;
+import org.elasticsearch.client.RequestConverters.Params;
+import org.elasticsearch.client.asyncsearch.SubmitAsyncSearchRequest;
+import org.elasticsearch.rest.action.search.RestSearchAction;
+
+import java.io.IOException;
+import java.util.Locale;
+
+import static org.elasticsearch.client.RequestConverters.REQUEST_BODY_CONTENT_TYPE;
+
+final class AsyncSearchRequestConverters {
+
+    static Request submitAsyncSearch(SubmitAsyncSearchRequest asyncSearchRequest) throws IOException {
+        String endpoint = new RequestConverters.EndpointBuilder().addCommaSeparatedPathParts(
+                asyncSearchRequest.getIndices())
+                .addPathPartAsIs("_async_search").build();
+        Request request = new Request(HttpPost.METHOD_NAME, endpoint);
+        Params params = new RequestConverters.Params();
+        // add all typical search params and search request source as body
+        addSearchRequestParams(params, asyncSearchRequest);
+        if (asyncSearchRequest.getSearchSource() != null) {
+            request.setEntity(RequestConverters.createEntity(asyncSearchRequest.getSearchSource(), REQUEST_BODY_CONTENT_TYPE));
+        }
+        // set async search submit specific parameters
+        if (asyncSearchRequest.isCleanOnCompletion() != null) {
+            params.putParam("clean_on_completion", asyncSearchRequest.isCleanOnCompletion().toString());
+        }
+        if (asyncSearchRequest.getKeepAlive() != null) {
+            params.putParam("keep_alive", asyncSearchRequest.getKeepAlive().getStringRep());
+        }
+        if (asyncSearchRequest.getWaitForCompletion() != null) {
+            params.putParam("wait_for_completion", asyncSearchRequest.getWaitForCompletion().getStringRep());
+        }
+        request.addParameters(params.asMap());
+        return request;
+    }
+
+    static void addSearchRequestParams(Params params, SubmitAsyncSearchRequest request) {
+        params.putParam(RestSearchAction.TYPED_KEYS_PARAM, "true");
+        params.withRouting(request.getRouting());
+        params.withPreference(request.getPreference());
+        params.withIndicesOptions(request.getIndicesOptions());
+        params.putParam("search_type", request.getSearchType().name().toLowerCase(Locale.ROOT));
+        params.putParam("max_concurrent_shard_requests", Integer.toString(request.getMaxConcurrentShardRequests()));
+        if (request.getRequestCache() != null) {
+            params.putParam("request_cache", Boolean.toString(request.getRequestCache()));
+        }
+        if (request.getAllowPartialSearchResults() != null) {
+            params.putParam("allow_partial_search_results", Boolean.toString(request.getAllowPartialSearchResults()));
+        }
+        params.putParam("batched_reduce_size", Integer.toString(request.getBatchedReduceSize()));
```

Comment:
nit: maybe given they are now used in two places we could add the with* methods for these params too?

Reply:
done in f65cf82